src/preset.ts

```diff
@@ -0,0 +1,1229 @@
+                          'Configure preset {0}: Specified cl.exe with toolset {1} and architecture {2} is not found, you may need to run "CMake: Scan for Compilers" if it exists on your computer.',
+                          preset.name, toolsetVsVersion ? `${toolsetVsVersion},${toolsetArch}` : toolsetArch, arch));
+          } else {
+            clEnv = getKitEnvironmentVariablesObject(await effectiveKitEnvironment(kits[latestVsIndex]));
+            // if ninja isn't on path, try to look for it in a VS install
+            const ninjaLoc = await execute('where.exe', ['ninja'], null, { environment: preset.environment as EnvironmentVariables,
+                                                                           silent: true,
+                                                                           encoding: 'utf8',
+                                                                           shell: true }).result;
+            if (!ninjaLoc.stdout) {
+              const vsCMakePaths = await paths.vsCMakePaths(kits[latestVsIndex].visualStudio);
+              if (vsCMakePaths.ninja) {
+                log.warning(localize('ninja.not.set', 'Ninja is not set on PATH, trying to use {0}', vsCMakePaths.ninja));
+                clEnv['PATH'] = `${path.dirname(vsCMakePaths.ninja)};${clEnv['PATH']}`;
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  clEnv = util.mergeEnvironment(inheritedEnv as EnvironmentVariables, clEnv as EnvironmentVariables);
+  preset.environment = util.mergeEnvironment(clEnv as EnvironmentVariables, preset.environment as EnvironmentVariables);
+
+  // Expand strings
+  const expansionOpts: ExpansionOptions = {
+    vars: {
+      generator: preset.generator || 'null',
+      workspaceFolder,
+      workspaceFolderBasename: path.basename(workspaceFolder),
+      workspaceHash: util.makeHashString(workspaceFolder),
+      workspaceRoot: workspaceFolder,
+      workspaceRootFolderName: path.dirname(workspaceFolder),
+      userHome: paths.userHome,
+      sourceDir,
+      sourceParentDir: path.dirname(sourceDir),
+      sourceDirName: path.basename(sourceDir),
+      presetName: preset.name
+    },
+    envOverride: preset.environment as EnvironmentVariables,
+    recursive: true,
+    // Don't support commands since expansion might be called on activation. If there is
+    // an extension depending on us, and there is a command in this extension is invoked,
+    // this would be a deadlock. This could be avoided but at a huge cost.
+    doNotSupportCommands: true
+  };
+  // Expand environment vars first since other fields may refer to them
+  if (preset.environment) {
+    for (const key in preset.environment) {
+      if (preset.environment[key]) {
+        preset.environment[key] = await expandString(preset.environment[key]!, expansionOpts);
+      }
+    }
+  }
+
+  removeNullEnvVars(preset);
+
+  // Expand other fields
+  if (preset.binaryDir) {
```

Comment:
The spec says that binaryDir must always be set unless it is inherited from another preset, but it is required to be set somewhere. I just got a crash when creating a new configure preset that just had a "name".
```
[rollbar] Unhandled exception: Failed writing to file .cmake\api\v1\query\client-vscode\query.json Error: EPERM: operation not permitted, mkdir 'C:\Program Files\Microsoft VS Code\.cmake' {}
```

Reply:
binaryDir

An optional string representing the path to the output binary directory. This field supports macro expansion. If a relative path is specified, it is calculated relative to the source directory. If binaryDir is not specified, it must be inherited from the inherits preset (unless this preset is hidden).

Reply:
I don't think this PR does the relative path support yet either.